lib/chef/resource/archive_file.rb

```diff
@@ -0,0 +1,133 @@
+
+      property :mode, [String, Integer],
+               description: "The mode of the extracted files",
+               default: "755"
+
+      property :destination, String,
+               description: "The file path to extract the archive file to.",
+               required: true
+
+      property :options, [Array, Symbol],
+               description: "An array of symbols representing extraction flags. Example: :no_overwrite to prevent overwriting files on disk.",
+               default: lazy { [] }
+
+      # backwards compatibility for the legacy cookbook names
+      alias_method :extract_options, :options
+      alias_method :extract_to, :destination
+
+      action :extract do
+        description "Extract and archive file."
+
+        unless ::File.exist?(new_resource.path)
+          raise Errno::ENOENT, "No archive found at #{new_resource.path}!"
+        end
+
+        unless Dir.exist?(new_resource.destination)
+          converge_by("create directory #{new_resource.destination}") do
+            FileUtils.mkdir_p(new_resource.destination, mode: new_resource.mode.to_i)
+          end
+        end
+
+        converge_by("extract #{new_resource.path} to #{new_resource.destination}") do
+          extract(new_resource.path, new_resource.destination,
+            Array(new_resource.options))
+        end
+
+        if new_resource.owner || new_resource.group
+          converge_by("set owner of #{new_resource.destination} to #{new_resource.owner}:#{new_resource.group}") do
+            FileUtils.chown_R(new_resource.owner, new_resource.group, new_resource.destination)
+          end
+        end
+      end
+
+      action_class do
+        # This can't be a constant since we might not have required 'ffi-libarchive' yet.
+        def extract_option_map
+          {
+            owner: Archive::EXTRACT_OWNER,
+            permissions: Archive::EXTRACT_PERM,
+            time: Archive::EXTRACT_TIME,
+            no_overwrite: Archive::EXTRACT_NO_OVERWRITE,
+            acl: Archive::EXTRACT_ACL,
+            fflags: Archive::EXTRACT_FFLAGS,
+            extended_information: Archive::EXTRACT_XATTR,
+            xattr: Archive::EXTRACT_XATTR,
+            no_overwrite_newer: Archive::EXTRACT_NO_OVERWRITE_NEWER,
+          }
+        end
+
+        # @param [String] src
+        # @param [String] dest
+        # @param [Array] options
+        #
+        # @return [Boolean] was the file extraction performed or not
+        def extract(src, dest, options = [])
+          require "ffi-libarchive"
+
+          flags = [options].flatten.map { |option| extract_option_map[option] }.compact.reduce(:|)
+          modified = false
+
+          Dir.chdir(dest) do
+            archive = Archive::Reader.open_filename(src)
+
+            archive.each_entry do |e|
+              pathname = ::File.expand_path(e.pathname)
+              if ::File.exist?(pathname)
+                modified = true unless ::File.mtime(pathname) == e.mtime
```

Comment:
should probably check file size here as well at least.

Reply:
oh wait or is size coming from the archive sometimes a pile of lies and that is why it isn't here?

Reply:
I'm not sure I'd trust that data